File: find-all-anagrams.java
// Time - O(M+N)
// Space - O(1)
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> result = new ArrayList<>();
        if(s == null || s.length() == 0) {
            return result;
        }
        
        HashMap<Character, Integer> map = new HashMap<>();
        
        
        for(int i=0;i<p.length();i++) {
            char ch = p.charAt(i);
            
            map.put(ch, map.getOrDefault(ch, 0) + 1);
        }
        int match = 0;
        for(int i=0;i<s.length();i++) {
            char ch = s.charAt(i);
            
            if(map.containsKey(ch)) {
                int cnt = map.get(ch);
                cnt--;
                if(cnt == 0) {
                    match++;
                }
                map.put(ch, cnt);
            }
            
            if(i>= p.length()) {
                char outgoing = s.charAt(i - p.length());
                if(map.containsKey(outgoing)) {
                    int cnt = map.get(outgoing);
                    cnt++;
                    if(cnt == 1) {
                        match--;
                    }
                    map.put(outgoing, cnt);
                }
            }
            
            if(match == map.size()) {
                result.add(i - p.length() + 1);
            }
        }
        
        return result;
    }
}

File: FindAllAnagrams.cs

// Time Complexity : O(m+n)
// Space Complexity : O(1) - hashampa contians max 26 charcters
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


public IList<int> FindAnagrams(string s, string p) {
    
    List<int> result = new List<int>();
    if(s == null || s.Length == 0 || p == null || p.Length == 0 || p.Length > s.Length)
        return result;
    
    Dictionary<char, int> pMap = new Dictionary<char, int>();
    foreach(char ch in p)
    {
        if(pMap.ContainsKey(ch))
            pMap[ch] += 1;
        else
            pMap.Add(ch, 1);
    }
    
    int match = 0;
    for(int i = 0; i < s.Length; i++)
    {
        //incoming character, if exists in hashmap, decrease the count and increase the meatch if count is 0
        char inChar = s[i];
        if(pMap.ContainsKey(inChar))
        {
            int cnt = pMap[inChar];
            cnt--;
            if(cnt == 0)                
                match++;                                    
                        
            pMap[inChar] = cnt;
        }
        
        //outgoing character, if exists in hasmap, increase the count in hasmap, if count becomes 1, derease match by 1
        if(i >= p.Length)
        {
            char outChar = s[i - p.Length];
            if(pMap.ContainsKey(outChar))
            {
                int cnt = pMap[outChar];
                cnt++;
                if(cnt == 1)
                    match--;
            
                pMap[outChar] = cnt;
            }
        }
        
        if(match == pMap.Count)
            result.Add(i - p.Length + 1);
    }
    
    return result;
}